### PR TITLE
Unified API

### PR DESCRIPTION
**Pull request by alexsu52**
Part 1: PR includes:
- Unified NNCF API implementation.
- The initial implementation of Unified API support for TF and Torch backends.

### CODE REVIEW

File: nncf/common/graph/model_transformer.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+class ModelTransformer:
+    """
+    Applies transformations to the model.
+    """
+
+    def __init__(self, model, transformation_layout):
```
Comment (andrey-churkin) [inline nncf/common/graph/model_transformer.py:19]: @alexsu52 Can we specify type hints?

File: nncf/common/graph/model_transformer.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        """
+        Initializes Model Transformer
+
+        :param model: The model to be transformed
+        :param transformation_layout: List of transformations
+        """
+        self._model = model
+        self._transformations = transformation_layout.transformations
```
Comment (andrey-churkin) [inline nncf/common/graph/model_transformer.py:27]: Looks like `transformation_layout` is not List. The docstring can be confusing.

File: nncf/compression_method_api.py [hunk 3]
```diff
@@ -106,58 +103,28 @@ def state_dict(self):
...
-        """
-        if self.value == other.value:
-            return self
-        return CompressionLevel.PARTIAL
-
-    def __lt__(self, other: 'CompressionLevel') -> bool:
-        return self.value < other.value
+    def __getattribute__(self, name):
+        if name == 'current_epoch':
+            return 0 if self._current_epoch == -1 else self._current_epoch
```
Comment (vshampor) [inline nncf/compression_method_api.py:108]: Why didn't `@property` work?

File: nncf/api/compression.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        parameters, and hooks inserted into the model to enable algorithm-specific
+        compression.
+
+        :param model: The model with additional modifications necessary to enable
+            algorithm-specific compression during fine-tuning.
+        :return: The instance of the `CompressionAlgorithmController`.
+        """
+        return CompressionAlgorithmController(model)
+
+    def get_transformation_layout(self, model: ModelType) -> TransformationLayout:
```
Comment (vshampor) [inline nncf/api/compression.py:307]: With this PR in place, NNCF for PyTorch still does not use `TransformationLayout`. The common API should not be extended with `TransformationLayout` at this point in time. It may be added later, but only along with the changes in the PyTorch part.

File: nncf/compression_method_api.py (hunk 3 above)
Comment (andrey-churkin) [reply nncf/compression_method_api.py:108]: Do we really need to return `0` when `_current_epoch == -1`? I think it is a legacy code. Probably it was needed when the method `epoch_step()` was called at the end of an epoch. But now, the scheduler should not do anything while the `epoch_step()` method will not be called. @vshampor Can you please confirm or reject this statement?

Comment (vshampor) [reply nncf/compression_method_api.py:108]: I, too, think that this code is no longer relevant. Would any of the tests fail if we removed it?

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/api/compression.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+    def __init__(self):
+        """
+        Initializes the internal state of the compression scheduler specified by:
+            - `current_step` is the index of the global training step, counted
+            from 0 to the end of training. The initial value is -1
+            - `current_epoch` is the training epoch index (numbering from zero).
+            The initial value is -1.
+
+        The `current_step` and` current_epoch` specify the training step and epoch,
```
Comment (andrey-churkin) [inline nncf/api/compression.py:95]: `current_step` and` current_epoch` --> `current_step` and `current_epoch`

File: nncf/api/compression.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        update the compression method state taking into account the training step,
+        there is the same for current_epoch is -1.
+        """
+        self.current_step = -1
+        self.current_epoch = -1
+
+    def step(self, next_step: Optional[int] = None) -> None:
+        """
+        Should be called at the beginning of each training step to prepare
+        the compression method to continue training the model in the 'next_step'.
```
Comment (andrey-churkin) [inline nncf/api/compression.py:111]: Please use ` instead of '

File: nncf/api/compression.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+            will update the state of the compression method.
+        """
+        if next_step is None:
+            next_step = self.current_step + 1
+        self.current_step = next_step
+
+    def epoch_step(self, next_epoch: Optional[int] = None) -> None:
+        """
+        Should be called at the beginning of each training epoch to prepare
+        the compression method to continue training the model in the 'next_epoch'.
```
Comment (andrey-churkin) [inline nncf/api/compression.py:123]: Please use ` instead of '

File: nncf/compression_method_api.py (hunk 3 above)
Comment (alexsu52) [reply nncf/compression_method_api.py:108]: Some tests fail if we removed it. I believe, that @andrey-churkin will remove it during work on the implementation of common schedulers.

Comment (andrey-churkin) [reply nncf/compression_method_api.py:108]: Yes, no problem.

File: nncf/api/compression.py (hunk 4 above)
Comment (alexsu52) [reply nncf/api/compression.py:307]: In line with your comment, I have opened a ticket

**Review by vshampor — APPROVED** [review]

**Comment by alexsu52** [comment]
Jenkins please retry a build

File: docs/styleguide/PyGuide.md [hunk 8]
```diff
@@ -526,23 +526,24 @@ Classes should have a docstring below the class definition describing the class.
...
         """
         Initializes ModelTransformer 
 
-        :param model: Keras model to be transformed
-        :param transformation_layout: list of transformations
+        :param model: The model to be transformed
+        :param transformation_layout: List of transformations
         """
         self._model = model
         self._transformation_layout = transformation_layout
```
Comment (andrey-churkin) [inline docs/styleguide/PyGuide.md:540]: Could you please update these lines?

Comment (alexsu52) [reply docs/styleguide/PyGuide.md:540]: Done. Please, check.

**Review by andrey-churkin — APPROVED** [review]

File: nncf/api/compression.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
+#
+#  Copyright (c) 2019-2020 Intel Corporation
```
Comment (AlexKoff88) [inline nncf/api/compression.py:2]: We have to update years in copyrights. I don't mind doing that in a different PR.

File: nncf/api/compression.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+extend the existing algorithms.
+"""
+from typing import Any, Dict, Optional, TypeVar
+
+from nncf import NNCFConfig
+from nncf.common.graph.model_transformer import ModelTransformer
+from nncf.common.graph.transformations.layout import TransformationLayout
+from nncf.common.utils.ordered_enum import OrderedEnum
+
+DOMAIN_CUSTOM_OPS_NAME = "org.openvinotoolkit"
```
Comment (AlexKoff88) [inline nncf/api/compression.py:26]: I am not sure about the current phase of unification but at some point `DOMAIN_CUSTOM_OPS_NAME ` should be moved to PyTorch-specific code.

File: nncf/api/compression.py (hunk 9 above)
Comment (alexsu52) [reply nncf/api/compression.py]: Done

File: nncf/api/compression.py (hunk 10 above)
Comment (alexsu52) [reply nncf/api/compression.py]: Rolled back the changes.
